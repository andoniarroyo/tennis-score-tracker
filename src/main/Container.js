// @flow
// external modules
import { connect } from 'react-redux';
// internal types
import type { StateType, MatchScoreType } from './index';
// internal modules
import scoreSelector from './selectors';
import TenisScoreTracker from './TennisScoreTracker';
import updateScore from './actions';

type StateToPropsType = {
    score: MatchScoreType
}

/**
 * Maps the application state to props passed to the component
 * @param {StateType} state The application redux state
 * @returns {StateToPropsType} The props coming from the state
 * @private
 */
const mapStateToProps = (state: StateType): StateToPropsType => ({
  score: scoreSelector(state),
});

const mapDispatchToProps = {
  updateScore,
};

export default connect(mapStateToProps, mapDispatchToProps)(TenisScoreTracker);
