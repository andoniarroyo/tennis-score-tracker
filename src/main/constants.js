// @flow
export const SETS_TO_WIN = 3;
export const MATCH_STATUS_INPLAY = 'inplay';
export const MATCH_STATUS_FINISHED = 'finished';
export const PLAYER1_NAME = 'player1';
export const PLAYER2_NAME = 'player2';

// actions related constants
export const UPDATE_SCORE_REQUESTED = 'UPDATE_SCORE_REQUESTED';
