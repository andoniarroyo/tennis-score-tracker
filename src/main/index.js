// @flow
import scoreReducer from './reducers';

export type PlayerScoreType = {
    sets: number,
    games: number,
    score: string,
    tieBreak: number,
    historicScoreboard: Array<number>,
}

export type MatchScoreType = {
    status: string,
    winner: string,
    player1: PlayerScoreType,
    player2: PlayerScoreType,
    inTiebreakMode: boolean,
}

export type StateType = {
    score: MatchScoreType
}

const reducers = {
  score: scoreReducer,
};

export default reducers;
