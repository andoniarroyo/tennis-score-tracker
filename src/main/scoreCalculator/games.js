// @flow
import type { UpdateSideEffectType, ScoreContextType } from '../matchScoreCalculator';

const MINIMUM_GAMES_TO_WIN_A_SET = 6;
const MINIMUM_GAMES_DIFFERENCE_TO_WIN_A_SET = 2;

export type NextGameSideEffectType = {
    isFinishingSet: boolean,
    updates: Array<UpdateSideEffectType>,
}

/**
 * Calculates the new games value for the winner of the point
 * @param {number} winnerGames Won games by the winner
 * @param {number} loserGames Won games by the loser
 * @return {number} The new won games by the winner of the point
 * @private
 */
const calculateWinnerGames = (winnerGames: number, loserGames: number): number => {
  const newWinnerGames = winnerGames + 1;
  return (newWinnerGames >= MINIMUM_GAMES_TO_WIN_A_SET &&
    newWinnerGames - loserGames >= MINIMUM_GAMES_DIFFERENCE_TO_WIN_A_SET) ? 0 : (newWinnerGames);
};

/**
 * Calculates the new games value for the loser of the point
 * @param {number} newWinnerGames Won games by the winner (including the current point)
 * @param {number} loserGames Won games by the loser
 * @return {number} The new won games by the loser of the point
 * @private
 */
const calculateLoserGames = (newWinnerGames: number, loserGames:number): number => (
  newWinnerGames === 0 ? 0 : loserGames
);

/**
 * Calculates if the point finishes the current set
 * @param {number} newWinnerGames Won games by the winner (including the current point)
 * @return {boolean} True if as a side effect of winning the point the current set ends
 * @private
 */
const calculateIsFinishingSet = (newWinnerGames: number): boolean => newWinnerGames === 0;

/**
 * Calculates if the match enters in tie break mode
 * @param {number} newWinnerGames Won games by the winner (including the current point)
 * @param {number} loserGames Won games by the loser
 * @return {boolean} True if the match enters in tie break mode, False otherwise
 * @private
 */
const calculateEnterTiebreakMode = (newWinnerGames: number, loserGames: number): boolean =>
  newWinnerGames === MINIMUM_GAMES_TO_WIN_A_SET && loserGames === MINIMUM_GAMES_TO_WIN_A_SET;

/**
  * Calculates the implications of win the point in the match score
  * @param {ScoreContextType} scoreContext The latest point context
  * @returns {NextGameSideEffectType} The side effects of the latest point in the match score
  * @public
  */
const calculateNextGame = ({ winnerName, loserName, matchScore }: ScoreContextType): NextGameSideEffectType => {
  const winnerGames = matchScore[winnerName].games;
  const loserGames = matchScore[loserName].games;

  const newWinnerGames = calculateWinnerGames(winnerGames, loserGames);
  const newLoserGames = calculateLoserGames(newWinnerGames, loserGames);
  const isFinishingSet = calculateIsFinishingSet(newWinnerGames);

  const newWinnerHistoricScoreboard = Array.from(matchScore[winnerName].historicScoreboard);
  const newLoserHistoricScoreboard = Array.from(matchScore[loserName].historicScoreboard);

  if (isFinishingSet) {
    newWinnerHistoricScoreboard.push(newWinnerGames || winnerGames + 1);
    newLoserHistoricScoreboard.push(loserGames);
  }

  const isInTiebreakMode = calculateEnterTiebreakMode(newWinnerGames, loserGames);

  return {
    isFinishingSet,
    updates: [
      {
        keyPath: `${winnerName},historicScoreboard`,
        value: newWinnerHistoricScoreboard,
      },
      {
        keyPath: `${loserName},historicScoreboard`,
        value: newLoserHistoricScoreboard,
      },
      {
        keyPath: 'inTiebreakMode',
        value: isInTiebreakMode,
      },
      {
        keyPath: `${winnerName},games`,
        value: newWinnerGames,
      },
      {
        keyPath: `${loserName},games`,
        value: newLoserGames,
      },
    ],
  };
};

export default calculateNextGame;
