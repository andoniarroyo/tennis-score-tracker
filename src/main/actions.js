// @flow
// internal types
import type { MatchScoreType } from './index';
// internal modules
import { UPDATE_SCORE_REQUESTED } from './constants';

export type UpdateScoreActionType = {
    type: string,
    newScore: MatchScoreType
};

/**
 * Action creator to update the match score
 * @param {MatchScoreType} newScore The new score value
 * @returns {UpdateScoreActionType} The generated action
 * @public
 */
const updateScore = (newScore: MatchScoreType): UpdateScoreActionType => (
  {
    type: UPDATE_SCORE_REQUESTED,
    newScore,
  }
);

export default updateScore;
