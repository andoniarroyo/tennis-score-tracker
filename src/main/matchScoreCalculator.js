// @flow
// internal types
import type { MatchScoreType } from './index';

// internal modules
import { PLAYER1_NAME, PLAYER2_NAME } from './constants';
import calculateNextPoint from './scoreCalculator/points';
import calculateNextGame from './scoreCalculator/games';
import calculateNextSet from './scoreCalculator/sets';
import calculateNextTieBreak from './scoreCalculator/tieBreak';

export type ScoreContextType = {
    winnerName: string,
    loserName: string,
    matchScore: MatchScoreType,
}
export type UpdateSideEffectType = {
    keyPath: string,
    value: mixed
}
export type UpdateSideEffectsType = Array<UpdateSideEffectType>;

/**
 * Gets the point loser player name
 * @param {string} winnerPlayer Name of the payer who won the point
 * @returns {string} The name of the loser player
 * @private
 */
const getLoserPlayer = (winnerPlayer: string): string => (
  winnerPlayer === PLAYER1_NAME ? PLAYER2_NAME : PLAYER1_NAME
);

/**
 * Clones the current match story
 * @param {MatchScoreType} previousMatchScore The match score object to be cloned
 * @returns {MatchScoreType} The new cloned score
 * @private
 */
const cloneMatchScore = (previousMatchScore: MatchScoreType): MatchScoreType => (
  Object.assign({}, previousMatchScore)
);

/**
 * Applies a side effect to the match score updating the state
 * @param {MatchScoreType} matchScore The match score to be used as reference
 * @param {UpdateSideEffectsType} updatesSideEffects The side effects to be applied
 * @returns {MatchScoreType} The match scrore resulting status
 * @private
 */
const applySideEffect = (matchScore: MatchScoreType, updatesSideEffects: UpdateSideEffectsType): MatchScoreType => {
  const newMatchScore = cloneMatchScore(matchScore);
  updatesSideEffects.forEach((update: UpdateSideEffectType) => {
    const keys = update.keyPath.split(',');
    if (keys.length === 1) {
      newMatchScore[keys[0]] = update.value;
      return newMatchScore;
    }
    let reference = matchScore;
    while (keys.length > 1) {
      reference = reference[keys.shift()];
    }
    reference[keys.shift()] = update.value;
    return reference;
  });
  return newMatchScore;
};

/**
 * Updates the incoming match score based on the player who won the point
 * @param {string} winnerName The name of the point winner
 * @param {MatchScoreType} previousMatchScore The match score to be used as reference
 * @returns {MatchScoreType} The new match score generated
 * @private
 */
const playerScored = (winnerName: string, previousMatchScore: MatchScoreType): MatchScoreType => {
  let matchScore = cloneMatchScore(previousMatchScore);
  const loserName = getLoserPlayer(winnerName);
  const scoreContext = { winnerName, loserName, matchScore };
  // Tiebreak mode has their special flow
  if (previousMatchScore.inTiebreakMode) {
    const newTiebreakSideEffect = calculateNextTieBreak(scoreContext);
    return applySideEffect(matchScore, newTiebreakSideEffect);
  }
  // score
  const newScoreSideEffect = calculateNextPoint(scoreContext);
  matchScore = applySideEffect(matchScore, newScoreSideEffect.updates);
  if (newScoreSideEffect.isFinishingGame) {
    // games
    const newGamesSideEffect = calculateNextGame(scoreContext);
    matchScore = applySideEffect(matchScore, newGamesSideEffect.updates);
    if (newGamesSideEffect.isFinishingSet) {
      // sets
      const newSetsSideEffect = calculateNextSet(scoreContext);
      matchScore = applySideEffect(matchScore, newSetsSideEffect);
    }
  }
  return matchScore;
};

/**
 * Updates the match score based on the player1 scored a point
 * @param {MatchScoreType} previousScore The previous match score
 * @returns {MatchScoreType} The new match score applying the side effects of the players 1 point
 * @public
 */
export const player1Scored = (previousScore: MatchScoreType): MatchScoreType =>
  playerScored(PLAYER1_NAME, previousScore);

/**
 * Updates the match score based on the player2 scored a point
 * @param {MatchScoreType} previousScore The previous match score
 * @returns {MatchScoreType} The new match score applying the side effects of the players 2 point
 * @public
 */
export const player2Scored = (previousScore: MatchScoreType): MatchScoreType =>
  playerScored(PLAYER2_NAME, previousScore);
