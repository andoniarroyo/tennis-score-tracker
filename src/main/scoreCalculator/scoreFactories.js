// @flow
// internal types
import type { PlayerScoreType, MatchScoreType } from '../index';
//internal modules
import { MATCH_STATUS_INPLAY } from '../constants';

/**
 * Composes a playscore with the values received by parameters
 * @param {number} sets The number of sets won by the player
 * @param {number} games The number of games won by the player
 * @param {string} score The number of points won by the player
 * @param {number} tieBreak The number of tiebreak points won by the player
 * @param {Array<number>} historicScoreboard The historic scoreboards for the player
 * @returns {PlayerScoreType} The player score generated object
 */
export const playerScoreFactory = (sets: number = 0, games: number = 0, score: string = '0',
  tieBreak: number = 0, historicScoreboard: Array<number> = []): PlayerScoreType => (
  {
    sets, games, score, tieBreak, historicScoreboard,
  }
);

/**
 * Composes a match score based on the received parameters
 * @param {PlayerScoreType} player1Score The player 1 score
 * @param {PlayerScoreType} player2Score The player 2 score
 * @param {boolean} inTiebreakMode The initial value for the in tiebreak mode
 * @returns {MatchScoreType} The match score generated object
 */
export const matchScoreFactory = (player1Score: PlayerScoreType = playerScoreFactory(),
  player2Score: PlayerScoreType = playerScoreFactory(), inTiebreakMode: boolean = false): MatchScoreType => (
  {
    status: MATCH_STATUS_INPLAY,
    winner: '',
    player1: player1Score,
    player2: player2Score,
    inTiebreakMode,
  }
);
