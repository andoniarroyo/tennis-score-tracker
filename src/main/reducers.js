// @flow
// internal types
import type { UpdateScoreActionType } from './actions';
import type { MatchScoreType } from './index';
// internal modules
import { matchScoreFactory } from './scoreCalculator/scoreFactories';
import { UPDATE_SCORE_REQUESTED } from './constants';

type CommonActionType = {
    type: string,
}

const reducers = {
  [UPDATE_SCORE_REQUESTED]:
        (score: MatchScoreType, { newScore }: UpdateScoreActionType) => newScore,
};

/**
 * Reduces the score in the state based on the received action
 * @param {object} score The current score
 * @param { CommonActionType } action The action to be evaluated
 * @returns {object} The new reduced score
 */
const scoreReducer = (score: MatchScoreType = matchScoreFactory(), action: CommonActionType) => {
  const reducer = reducers[action.type];
  return reducer ? reducer(score, action) : score;
};

export default scoreReducer;
