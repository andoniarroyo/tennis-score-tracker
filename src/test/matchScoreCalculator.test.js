import { SETS_TO_WIN } from '../main/constants';
import { player1Scored, player2Scored } from '../main/matchScoreCalculator';
import { matchScoreFactory, playerScoreFactory } from '../main/scoreCalculator/scoreFactories';

describe('The matchScoreCalculator', () => {
  it('increases the player 1 score when the player1 scored', () => {
    const previousMatchScore = matchScoreFactory();
    const newMatchScore = player1Scored(previousMatchScore);

    expect(newMatchScore).toMatchSnapshot();
  });

  it('increases the player 2 score when the player1 scored', () => {
    const previousMatchScore = matchScoreFactory();
    const newMatchScore = player2Scored(previousMatchScore);

    expect(newMatchScore).toMatchSnapshot();
  });

  describe('scoring without change the game', () => {
    it('increases the score from 0 to 15', () => {
      const previousMatchScore = matchScoreFactory();
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
    it('increases the score from 15 to 30', () => {
      const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 0, '15', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
    it('increases the score from 30 to 40', () => {
      const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 0, '30', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
    it('gives advantage if both players were in deuce (40 - 40)', () => {
      const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 0, '40', 0),
        playerScoreFactory(0, 0, '40', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
    it('sets the deuce again if the other player had advantage', () => {
      const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 0, '40', 0),
        playerScoreFactory(0, 0, 'A', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
  });

  describe('increasing the game', () => {
    it('increases the game when the player wins the game', () => {
      const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 0, '40', 0),
        playerScoreFactory(0, 1, '30', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
    it('increases the game when the player wins the game from the advantage', () => {
      const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 0, 'A', 0),
        playerScoreFactory(0, 0, '40', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
    it('increases the sets when the player wins 6 games and the other player 4 or less', () => {
      const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 5, '40', 0),
        playerScoreFactory(0, 4, '30', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
    it(`in spite of reaching 6 games, does not increase the sets if the difference is not equal 
        or bigger than 2`, () => {
        const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 5, '40', 0),
          playerScoreFactory(0, 5, '30', 0));
        const newMatchScore = player1Scored(previousMatchScore);

        expect(newMatchScore).toMatchSnapshot();
      });
    it('in case both players win 6 games the tie break mode starts', () => {
      const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 5, '40', 0),
        playerScoreFactory(0, 6, '30', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });

    describe('in tie break mode', () => {
      it('adds point to the tiebreak', () => {
        const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 6, '0', 0),
          playerScoreFactory(0, 6, '0', 0), true);
        const newMatchScore = player1Scored(previousMatchScore);

        expect(newMatchScore).toMatchSnapshot();
      });
      it('continue adding points if the difference is smaller than 2 points', () => {
        const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 6, '0', 7),
          playerScoreFactory(0, 6, '0', 7), true);
        const newMatchScore = player1Scored(previousMatchScore);

        expect(newMatchScore).toMatchSnapshot();
      });
      it(`wins the game and resolves the tiebreak mode if the value is 7 or more and the difference 
          is equal or bigger than  2`, () => {
          const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 6, '0', 7),
            playerScoreFactory(0, 6, '0', 6), true);
          const newMatchScore = player1Scored(previousMatchScore);

          expect(newMatchScore).toMatchSnapshot();
        });
    });
  });
  describe('increasing the sets', () => {
    it(`increases the sets for the scoring player and reset the games and scores 
        for both players`, () => {
        const previousMatchScore = matchScoreFactory(playerScoreFactory(0, 5, '40', 0),
          playerScoreFactory(0, 4, '30', 0));
        const newMatchScore = player1Scored(previousMatchScore);

        expect(newMatchScore).toMatchSnapshot();
      });
    it('finishes the match when a player wins the defined set to win amount of sets', () => {
      const currentWonSets = SETS_TO_WIN - 1;
      const previousMatchScore = matchScoreFactory(playerScoreFactory(currentWonSets, 5, '40', 0),
        playerScoreFactory(currentWonSets, 4, '30', 0));
      const newMatchScore = player1Scored(previousMatchScore);

      expect(newMatchScore).toMatchSnapshot();
    });
  });
});
