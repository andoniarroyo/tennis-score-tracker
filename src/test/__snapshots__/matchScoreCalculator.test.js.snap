// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The matchScoreCalculator increases the player 1 score when the player1 scored 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "15",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increases the player 2 score when the player1 scored 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "15",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the game in case both players win 6 games the tie break mode starts 1`] = `
Object {
  "inTiebreakMode": true,
  "player1": Object {
    "games": 6,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 6,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the game in spite of reaching 6 games, does not increase the sets if the difference is not equal 
        or bigger than 2 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 6,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 5,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the game in tie break mode adds point to the tiebreak 1`] = `
Object {
  "inTiebreakMode": true,
  "player1": Object {
    "games": 6,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 1,
  },
  "player2": Object {
    "games": 6,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the game in tie break mode continue adding points if the difference is smaller than 2 points 1`] = `
Object {
  "inTiebreakMode": true,
  "player1": Object {
    "games": 6,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 8,
  },
  "player2": Object {
    "games": 6,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 7,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the game in tie break mode wins the game and resolves the tiebreak mode if the value is 7 or more and the difference 
          is equal or bigger than  2 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [
      7,
    ],
    "score": "0",
    "sets": 1,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [
      6,
    ],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the game increases the game when the player wins the game 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 1,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 1,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the game increases the game when the player wins the game from the advantage 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 1,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the game increases the sets when the player wins 6 games and the other player 4 or less 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [
      6,
    ],
    "score": "0",
    "sets": 1,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [
      4,
    ],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator increasing the sets finishes the match when a player wins the defined set to win amount of sets 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [
      6,
    ],
    "score": "0",
    "sets": 3,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [
      4,
    ],
    "score": "0",
    "sets": 2,
    "tieBreak": 0,
  },
  "status": "finished",
  "winner": "player1",
}
`;

exports[`The matchScoreCalculator increasing the sets increases the sets for the scoring player and reset the games and scores 
        for both players 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [
      6,
    ],
    "score": "0",
    "sets": 1,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [
      4,
    ],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator scoring without change the game gives advantage if both players were in deuce (40 - 40) 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "A",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "40",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator scoring without change the game increases the score from 0 to 15 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "15",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator scoring without change the game increases the score from 15 to 30 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "30",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator scoring without change the game increases the score from 30 to 40 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "40",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "0",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;

exports[`The matchScoreCalculator scoring without change the game sets the deuce again if the other player had advantage 1`] = `
Object {
  "inTiebreakMode": false,
  "player1": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "40",
    "sets": 0,
    "tieBreak": 0,
  },
  "player2": Object {
    "games": 0,
    "historicScoreboard": Array [],
    "score": "40",
    "sets": 0,
    "tieBreak": 0,
  },
  "status": "inplay",
  "winner": "",
}
`;
